h5py/_hl/attrs.py

```diff
@@ -95,13 +96,32 @@ def get_id(self, name):
 
     @with_phil
     def __setitem__(self, name, value):
-        """ Set a new attribute, overwriting any existing attribute.
+        """ Set a new attribute, overwriting or modifying any existing attribute.
+
+        Overwrite existing attributes if track_order is False, otherwise modify
+        attribute if type is preserved.
 
         The type and shape of the attribute are determined from the data.  To
         use a specific type or shape, or to preserve the type of an attribute,
         use the methods create() and modify().
         """
-        self.create(name, data=value)
+        with phil:
+            # see issue
+            # if attribute exists and track_order=True
+            if name in self and self._id.get_create_plist().get_attr_creation_order() > 0:
+                # extract type
+                if not isinstance(value, Empty):
+                    value = base.array_for_new_object(value, specified_dtype=None)
+                dtype = value.dtype
+                attr = h5a.open(self._id, self._e(name))
+                dt = attr.dtype
+                attr.close()
+                # if type is preserved -> modify
+                if dtype == dt:
```

Comment:
I don't think this is quite specific enough, because it ignores dtype metadata:

```
In [8]: h5py.string_dtype('ascii') == h5py.string_dtype('utf-8')
Out[8]: True
```

I thought comparing the converted HDF5 types would be better, but it still seems to ignore the character set information. :confused: Perhaps we could check `and dtype.metadata == dt.metadata`.

Reply:
@takluyver Thanks for the suggestion. I'll have a look into this.